common/logger/src/libra_logger.rs

```diff
@@ -0,0 +1,286 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    counters::{
+        PROCESSED_STRUCT_LOG_COUNT, SENT_STRUCT_LOG_COUNT, STRUCT_LOG_PARSE_ERROR_COUNT,
+        STRUCT_LOG_QUEUE_ERROR_COUNT, STRUCT_LOG_SEND_ERROR_COUNT,
+    },
+    struct_log::TcpWriter,
+    Level, StructuredLogEntry,
+};
+use std::{
+    env, fmt,
+    io::Write,
+    sync::{
+        mpsc::{self, Receiver, SyncSender},
+        Arc,
+    },
+    thread,
+};
+
+pub const CHANNEL_SIZE: usize = 10000; // MAX_LOG_LINE_SIZE * 10000 = max buffer size
+const NUM_SEND_RETRIES: u8 = 1;
+
+pub struct LibraLoggerBuilder {
+    channel_size: usize,
+    level: Level,
+    address: Option<String>,
+    printer: Option<Box<dyn Writer>>,
+    is_async: bool,
+}
+
+impl LibraLoggerBuilder {
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            channel_size: CHANNEL_SIZE,
+            level: Level::Debug,
+            address: None,
+            printer: Some(Box::new(StderrWriter)),
+            is_async: false,
+        }
+    }
+
+    pub fn address(&mut self, address: String) -> &mut Self {
+        self.address = Some(address);
+        self
+    }
+
+    pub fn read_env(&mut self) -> &mut Self {
+        if let Ok(address) = env::var("STRUCT_LOG_TCP_ADDR") {
+            self.address(address);
+        }
+        self
+    }
+
+    pub fn level(&mut self, level: Level) -> &mut Self {
+        self.level = level;
+        self
+    }
+
+    pub fn channel_size(&mut self, channel_size: usize) -> &mut Self {
+        self.channel_size = channel_size;
+        self
+    }
+
+    pub fn printer(&mut self, printer: Box<dyn Writer + Send + Sync + 'static>) -> &mut Self {
+        self.printer = Some(printer);
+        self
+    }
+
+    pub fn is_async(&mut self, is_async: bool) -> &mut Self {
+        self.is_async = is_async;
+        self
+    }
+
+    pub fn init(&mut self) {
+        self.build()
+    }
+
+    pub fn build(&mut self) {
+        let logger = if self.is_async {
+            let (sender, receiver) = mpsc::sync_channel(self.channel_size);
+            let service = LoggerService {
+                receiver,
+                address: self.address.clone(),
+                printer: self.printer.take(),
+            };
+            let logger = Arc::new(LibraLogger {
+                sender: Some(sender),
+                printer: None,
+                level: self.level,
+            });
+
+            thread::spawn(move || service.run());
+            logger
+        } else {
+            Arc::new(LibraLogger {
+                sender: None,
+                printer: self.printer.take(),
+                level: self.level,
+            })
+        };
+
+        crate::logger::set_global_logger(logger.clone());
+        let logger = Box::leak(Box::new(logger));
+        crate::struct_log::set_struct_logger(logger).expect("Unable to setup structured logger");
```

Comment:
Unfortunately this expect causes some binaries to crash ... cli, is it really that disastrous if someone calls this twice?